mutation_writer/shard_based_splitting_writer.cc

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mutation_writer/shard_based_splitting_writer.hh"
+
+#include <cinttypes>
+#include <boost/range/adaptor/transformed.hpp>
+#include <boost/range/algorithm/min_element.hpp>
+#include <seastar/core/shared_mutex.hh>
+
+#include "mutation_reader.hh"
+
+namespace mutation_writer {
+
+class shard_based_splitting_mutation_writer {
```

Comment:
@denesb couldn't multishard_writer be parametrized instead?

I don't think it's trivial, but it's really screaming that we should be able to drop the make_foreign_reader() somewhere and get a fully-local splitter.

Reply:
Not only it could, I wrote as much in my introduction.
I however think this splitter is very simple, and changing the multishard_writer is both complex and risky. It becomes even more risky if done by me, who don't understand it that well.